### PR TITLE
Fix video backward seeking / stepping back sometimes getting stuck (in the presence of b-frames)

### PR DESCRIPTION
**Pull request by Wumpf**
### What

* Fixes https://github.com/rerun-io/rerun/issues/7956

Our PTS -> sample search algorithm wasn't quite right. Fixed this here and added a unit test.
Unfortunately, I had to introduce a small auxiliary data structure to keep this snappy / not O(n) for n samples, but it's very simple and rather contained.

Tested on native & web. Web still suffers from
* https://github.com/rerun-io/rerun/issues/7961

but on Chrome this works fine now!

https://github.com/user-attachments/assets/7a20467d-4d66-456f-be99-debba4ff7a31

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Using examples from latest `main` build: [rerun.io/viewer](https://rerun.io/viewer/pr/8053?manifest_url=https://app.rerun.io/version/main/examples_manifest.json)
  * Using full set of examples from `nightly` build: [rerun.io/viewer](https://rerun.io/viewer/pr/8053?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG
* [x] If applicable, add a new check to the [release checklist](https://github.com/rerun-io/rerun/blob/main/tests/python/release_checklist)!
* [x] If have noted any breaking changes to the log API in `CHANGELOG.md` and the migration guide

- [PR Build Summary](https://build.rerun.io/pr/8053)
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

To run all checks from `main`, comment on the PR with `@rerun-bot full-check`.

### CODE REVIEW

File: crates/store/re_video/src/demux/mod.rs [hunk 1]
```diff
@@ -89,6 +89,11 @@ pub struct SamplesStatistics {
     ///
     /// If true, the video typically has no B-frames as those require frame reordering.
     pub dts_always_equal_pts: bool,
+
+    /// If `dts_always_equal_pts` is false, then this gives for each sample whether its PTS is the highest seen so far.
+    /// If `dts_always_equal_pts` is true, then this is left empty.
+    /// This is used for optimizing PTS search.
+    pub has_sample_highest_pts_so_far: Vec<bool>,
```
Comment (jprochazk) [inline crates/store/re_video/src/demux/mod.rs:96]: Would it make sense to use `bit-vec` instead here?

File: crates/store/re_video/src/demux/mod.rs [hunk 2]
```diff
@@ -104,9 +109,23 @@ impl SamplesStatistics {
             .iter()
             .all(|s| s.decode_timestamp == s.presentation_timestamp);
 
+        let mut biggest_pts_so_far = Time::MIN;
+        let has_sample_highest_pts_so_far = samples
```
Comment (jprochazk) [inline crates/store/re_video/src/demux/mod.rs:113]: I'm also a bit worried about how many times we're iterating over the samples here. I don't think LLVM will optimize this to one loop.

**Review by jprochazk — APPROVED** [review]
Other than nits above this LGTM

File: crates/store/re_video/src/demux/mod.rs (hunk 1 above)
Comment (Wumpf) [reply crates/store/re_video/src/demux/mod.rs:96]: ah, yes good call!

File: crates/store/re_video/src/demux/mod.rs (hunk 2 above)
Comment (Wumpf) [reply crates/store/re_video/src/demux/mod.rs:113]: tbh I'm not sure if it's better to combine everything to a single loop or to loop several times: individual loops might vectorize better

pure speculation until someone actually measures. Maybe I should. After all we might run this with _a lot_ of samples!

Comment (jprochazk) [reply crates/store/re_video/src/demux/mod.rs:113]: It optimizes a lot better than I thought: https://godbolt.org/z/n3Yvzff9e

Comment (Wumpf) [reply crates/store/re_video/src/demux/mod.rs:113]: neither is in the mood for simd though. probably because everything is 64bit integer

Comment (Wumpf) [reply crates/store/re_video/src/demux/mod.rs]: leaving it be as is for now

**Comment by Wumpf** [comment]
known issue on `error[unmaintained]: `instant` is unmaintained`

**Pull request merged**